add fixes to run regression tests in Docker container

red1452 opened the pull request:
### Run regression tests in Docker container

You can build your Docker image from GPDB source or use prebuilt images from hub.adsw.io.
How to build Docker image: (["readme.md"](https://github.com/arenadata/gpdb/blob/f7ff7c8ecae4ce7ab3b73fd46171cdaa457b3591/arenadata/readme.md)).

1. Download the cmake-3.20 install script from ([source](https://github.com/Kitware/CMake/releases/download/v3.20.0/cmake-3.20.0-linux-x86_64.sh)).

2. Build diskquota in the Docker container.
Change <PATH_TO_DISKQUOTA_SRC> and <PATH_TO_CMAKE_INSTALL_SCRIPT> to the appropriate paths on your local machine.

```
docker run --rm -it -e DISKQUOTA_OS=rhel7 \
       -v /tmp/diskquota_artifacts:/home/gpadmin/diskquota_artifacts \
       -v <PATH_TO_DISKQUOTA_SRC>:/home/gpadmin/diskquota_src \
       -v <PATH_TO_CMAKE_INSTALL_SCRIPT>:/home/gpadmin/bin_cmake/cmake-3.20.0-linux-x86_64.sh \
       hub.adsw.io/library/gpdb6_regress:latest diskquota_src/concourse/scripts/entry.sh build
```

3. Run tests.
Change <PATH_TO_DISKQUOTA_SRC> and <PATH_TO_CMAKE_INSTALL_SCRIPT> to the appropriate paths on your local machine.

```
docker run --rm -it --sysctl 'kernel.sem=500 1024000 200 4096' \
       -v /tmp/diskquota_artifacts:/home/gpadmin/bin_diskquota \
       -v <PATH_TO_DISKQUOTA_SRC>:/home/gpadmin/diskquota_src \
       -v <PATH_TO_CMAKE_INSTALL_SCRIPT>:/home/gpadmin/bin_cmake/cmake-3.20.0-linux-x86_64.sh \
       hub.adsw.io/library/gpdb6_regress:latest diskquota_src/concourse/scripts/entry.sh test
```

red1452 committed:
add fixes to run regression tests in Docker container (#14)

Stolb27 commented on concourse/scripts/test_diskquota.sh:
```diff
@@ -18,6 +18,9 @@ function _main() {
     tar -xzf /home/gpadmin/bin_diskquota/diskquota-*-*.tar.gz -C /usr/local/greenplum-db-devel
     source /home/gpadmin/gpdb_src/gpAux/gpdemo/gpdemo-env.sh
 
+    gpconfig -c shared_preload_libraries -v "diskquota-2.2"
+    gpstop -ar
+
```
@red1452, Why this line here isn't work as expected? 

https://github.com/arenadata/diskquota/blob/df260b2d583b6136d1b8fe1c6982d89127f058b2/tests/regress/sql/config.sql#L4

Where does test failed with error without lines above?

red1452 replied:
Without my fixes:
```
+ cmake --build . --target installcheck
Built target install_test_extension
(using postmaster on Unix socket, port 6000)
============== dropping database "contrib_regression" ==============
NOTICE:  database "contrib_regression" does not exist, skipping
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing gp_inject_fault             ==============
CREATE EXTENSION
============== installing diskquota_test              ==============
ERROR:  [diskquota] booting 2.2.0, but diskquota-2.2 not in shared_preload_libraries. abort. (diskquota.c:178)
command failed: "/usr/local/greenplum-db-devel/bin/psql" -X -c "CREATE EXTENSION IF NOT EXISTS \"diskquota_test\"" "contrib_regression"
======================================================================
DIFF FILE: /home/gpadmin/diskquota_artifacts/tests/regress/regression.diffs
======================================================================


gmake[3]: *** [tests/CMakeFiles/regress] Error 1
gmake[2]: *** [tests/CMakeFiles/regress.dir/all] Error 2
gmake[1]: *** [tests/CMakeFiles/installcheck.dir/rule] Error 2
gmake: *** [installcheck] Error 2
[root@5be6c7bf072f gpadmin]# su - gpadmin
Last login: Wed Apr 19 10:07:02 UTC 2023 on pts/0
-bash-4.2$ source /usr/local/greenplum-db-devel/greenplum_path.sh;
-bash-4.2$ source gpdb_src/gpAux/gpdemo/gpdemo-env.sh;
-bash-4.2$ gpconfig -s shared_preload_libraries
Values on all segments are consistent
GUC          : shared_preload_libraries
Master  value: 
Segment value: 
-bash-4.2$ exit
```

It does not work because before execute `config.sql` extension `diskquota_test` is configured. Create extension `disquota_test` requires to create extension `diskquota`:
```
postgres=# CREATE EXTENSION IF NOT EXISTS diskquota_test;
ERROR:  [diskquota] booting 2.2.0, but diskquota-2.2 not in shared_preload_libraries. abort. (diskquota.c:178)
```

Loading extension `diskquota_test` is set at `CMakeLists.txt` for `RegressTarget_Add(regress`:
```
RegressTarget_Add(regress
    INIT_FILE
    ${CMAKE_CURRENT_SOURCE_DIR}/init_file
    SQL_DIR ${CMAKE_CURRENT_SOURCE_DIR}/regress/sql
    EXPECTED_DIR ${CMAKE_CURRENT_SOURCE_DIR}/regress/expected${EXPECTED_DIR_SUFFIX}
    RESULTS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/regress/results
    DATA_DIR ${CMAKE_CURRENT_SOURCE_DIR}/data
    SCHEDULE_FILE ${CMAKE_CURRENT_SOURCE_DIR}/regress/diskquota_schedule${EXPECTED_DIR_SUFFIX}
    REGRESS_OPTS
    --load-extension=gp_inject_fault
    --load-extension=diskquota_test
    --dbname=contrib_regression)
```

red1452 replied:
Command `gpconfig -c shared_preload_libraries -v $(./data/current_binary_name);` works correctly:
```
-bash-4.2$ gpconfig -c shared_preload_libraries -v $(./data/current_binary_name);
20230419:10:37:16:012695 gpconfig:5be6c7bf072f:gpadmin-[INFO]:-completed successfully with parameters '-c shared_preload_libraries -v diskquota-2.2.so'
-bash-4.2$ gpstop -raf
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Starting gpstop with args: -raf
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Gathering information and validating the environment...
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Obtaining Greenplum Master catalog information
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Obtaining Segment details from master...
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Greenplum Version: 'postgres (Greenplum Database) 6.22.1_arenadata41+dev.307.g60673e0 build commit:60673e0b66bf5af3621a641acef88c9059321870'
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Commencing Master instance shutdown with mode='fast'
20230419:10:37:23:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Master segment instance directory=/home/gpadmin/gpdb_src/gpAux/gpdemo/datadirs/qddir/demoDataDir-1
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Attempting forceful termination of any leftover master process
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Terminating processes for segment /home/gpadmin/gpdb_src/gpAux/gpdemo/datadirs/qddir/demoDataDir-1
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Stopping master standby host 5be6c7bf072f mode=fast
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Successfully shutdown standby process on 5be6c7bf072f
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Targeting dbid [2, 5, 3, 6, 4, 7] for shutdown
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Commencing parallel primary segment instance shutdown, please wait...
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-0.00% of jobs completed
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-100.00% of jobs completed
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Commencing parallel mirror segment instance shutdown, please wait...
20230419:10:37:24:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-0.00% of jobs completed
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-100.00% of jobs completed
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-----------------------------------------------------
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-   Segments stopped successfully      = 6
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-   Segments with errors during stop   = 0
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-----------------------------------------------------
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Successfully shutdown 6 of 6 segment instances 
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Database successfully shutdown with no errors reported
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Cleaning up leftover gpmmon process
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-No leftover gpmmon process found
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Cleaning up leftover gpsmon processes
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-No leftover gpsmon processes on some hosts. not attempting forceful termination on these hosts
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Cleaning up leftover shared memory
20230419:10:37:25:012833 gpstop:5be6c7bf072f:gpadmin-[INFO]:-Restarting System...
-bash-4.2$ psql postgres
psql (9.4.26)
Type "help" for help.

postgres=# CREATE EXTENSION IF NOT EXISTS diskquota_test;
CREATE EXTENSION
postgres=# create extension diskquota;
CREATE EXTENSION
postgres=# 
```

RekGRpth replied:
> Where does test failed with error without lines above?

IMHO, command

> --load-extension=diskquota_test

does only 

> CREATE EXTENSION IF NOT EXISTS \"diskquota_test\"

and does it before executing 

> diskquota/tests/regress/sql/config.sql

Stolb27 replied:
@RekGRpth, @red1452, how creating of diskquota_test leads to creating/initializing diskquota extension? At the first glance, this extension contains only functions definition.

Stolb27 replied:
As far I understand, there are C-functions in the diskquota_test. Creating of these functions lead to loading shared library and calling PG_init function that throws the given error. But how it works in the upstream CI?!

Stolb27 replied:
I have assumption about a possible reason. There is the next line in cmake files: https://github.com/arenadata/diskquota/blob/e4a06ea5a1b8a38eae2c194be11643f48dc8d39e/tests/CMakeLists.txt#L45

I didn't find any guaranties about processing order of dependencies. But if we run isolation tests before regression, all tests pass successfully:

```diff
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 9a77457..904ae8d 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -41,7 +41,7 @@ add_custom_target(install_test_extension

 add_custom_target(installcheck)
 add_dependencies(isolation2 install_test_extension)
-add_dependencies(regress install_test_extension)
+add_dependencies(regress install_test_extension isolation2)
 add_dependencies(installcheck isolation2 regress)
 ```

Anyway, I don't think that it's a better solution...

RekGRpth replied:
> if we run isolation tests before regression, all tests pass successfully

it is because `isolation2` test has not `--load-extension=diskquota_test` while in `config.sql` it adds `diskquota` to `shared_preload_libraries`

red1452 replied:
I remove extra dependence `--load-extension=diskquota_test` and all tests are passed:
```
Built target install_test_extension
(using postmaster on Unix socket, port 6000)
============== dropping database "contrib_regression" ==============
NOTICE:  database "contrib_regression" does not exist, skipping
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing gp_inject_fault             ==============
CREATE EXTENSION
============== checking optimizer status              ==============
Optimizer enabled. Using optimizer answer files whenever possible
============== checking gp_resource_manager status    ==============
Resource group disabled. Using default answer files
============== running regression test queries        ==============
test config                   ... ok (7.33 sec)  (diff:0.05 sec)
test test_create_extension    ... ok (1.98 sec)  (diff:0.05 sec)
...
test reset_config             ... ok (1.73 sec)  (diff:0.05 sec)

======================
 All 60 tests passed. 
======================
```
At tests SQL files where functions from diskquota_test are used, at first does `CREATE EXTENSION diskquota_test;`.
It means, that no need to do it before all tests are started.

Stolb27 commented on README.md:
````diff
@@ -317,3 +317,33 @@ schemas or roles which quota limit are reached.
+       -v <PATH_TO_DISKQUOTA>:/home/gpadmin/diskquota_src \
+       -v <PATH_TO_CMAKE_INSTALL_SCRIPT>:/home/gpadmin/bin_cmake/cmake-3.20.0-linux-x86_64.sh \
+       adb-6.x-dev:latest bash -c "/home/gpadmin/diskquota_src/concourse/scripts/entry.sh test"
+```
````
I think that this chapter may be moved to a separate arenadata/README.md file to avoid conflicts.

red1452 replied:
Done.

red1452 committed:
move help to new file

Stolb27 commented on arenadata/README.md:
````diff
@@ -0,0 +1,30 @@
+
+1. Build Docker container with GPDB from source directory:
+```
+docker build -t "adb-6.x-dev:latest" -f arenadata/Dockerfile .
````
Do we really need to rebuild gpdb container each time? Why not reuse `hub.adsw.io/library/gpdb6_regress:latest` in most cases from our registry? But you may provide a link to the appropriate article from gpdb repo here.

red1452 committed:
readme fixes and remove extra dependences

RekGRpth commented on concourse/scripts/entry.sh:
```diff
@@ -130,11 +127,6 @@ function create_fake_gpdb_src() {
         grep -rhw '/usr/local/greenplum-db-devel' -e 'abs_top_srcdir = .*' |\
         head -n 1 | awk '{ print $NF; }')"
 
-    if [ -d "${fake_gpdb_src}" ]; then
```
Why these lines were removed?

red1452 replied:
This check works incorrect. Command 
```
$(\
        grep -rhw '/usr/local/greenplum-db-devel' -e 'abs_top_srcdir = .*' |\
        head -n 1 | awk '{ print $NF; }')
```
returns string `/home/gpadmin/gpdb_src`.
This path is path of sources of GPDB. This directory is not empty. Sources are not configured. But this check looks: if directory is not empty than configure is not needed and skip this step.

Stolb27 replied:
Ok. What does do code block below (lines after configure block)?

red1452 replied:
These lines do useless things:
```
+ local fake_root
++ dirname /home/gpadmin/gpdb_src
+ fake_root=/home/gpadmin
+ mkdir -p /home/gpadmin
+ ln -s /home/gpadmin/gpdb_src /home/gpadmin/gpdb_src
```
At our countainer there is directory `/home/gpadmin/gpdb_src` and these commands do nothing.

Stolb27 commented on arenadata/README.md:
```diff
@@ -0,0 +1,28 @@
+       -v /tmp/diskquota_artifacts:/home/gpadmin/diskquota_artifacts \
+       -v <PATH_TO_DISKQUOTA_SRC>:/home/gpadmin/diskquota_src \
+       -v <PATH_TO_CMAKE_INSTALL_SCRIPT>:/home/gpadmin/bin_cmake/cmake-3.20.0-linux-x86_64.sh \
+       hub.adsw.io/library/gpdb6_regress:latest -c "/home/gpadmin/diskquota_src/concourse/scripts/entry.sh build"
```
You replaced bash in command line. It seems that absolute path is redundant it all.

```suggestion
       hub.adsw.io/library/gpdb6_regress:latest diskquota_src/concourse/scripts/entry.sh build
```

Stolb27 commented on arenadata/README.md:
```diff
@@ -0,0 +1,28 @@
+       -v /tmp/diskquota_artifacts:/home/gpadmin/bin_diskquota \
+       -v <PATH_TO_DISKQUOTA_SRC>:/home/gpadmin/diskquota_src \
+       -v <PATH_TO_CMAKE_INSTALL_SCRIPT>:/home/gpadmin/bin_cmake/cmake-3.20.0-linux-x86_64.sh \
+       hub.adsw.io/library/gpdb6_regress:latest -c "/home/gpadmin/diskquota_src/concourse/scripts/entry.sh test"
```
same issue

```suggestion
       hub.adsw.io/library/gpdb6_regress:latest diskquota_src/concourse/scripts/entry.sh test
```

Stolb27 commented on arenadata/README.md:
```diff
@@ -0,0 +1,28 @@
+# Run regression tests in Docker container
+
+You can build your Docker image from GPDB source or use already builded images from hub.adsw.io.
+How to build Docker image: https://github.com/arenadata/gpdb/blob/f7ff7c8ecae4ce7ab3b73fd46171cdaa457b3591/arenadata/readme.md
```
Please inline ALL links ([example](https://github.com/arenadata/gpdb/blob/f7ff7c8ecae4ce7ab3b73fd46171cdaa457b3591/arenadata/readme.md))

red1452 committed:
fixes

red1452 commented:
Commit message:
```
add fixes to run regression tests in Docker container (#14)

Removed useless commands from `concourse/scripts/entry.sh`, because our
docker images already have source of GPDB at `/home/gpadmin/gpdb_src`
and do not need to create or configure this directory.
Added file `arenadata/README.md` with information about how to run
tests at docker container.
Removed `--load-extension=diskquota_test` from `CMakeLists.txt` at regression
tests. Creating extension `diskquota_test` requires configured extension `diskquota`.
But `diskquota` is configured when `tests/regress/sql/config.sql` is executed, and
load-extension is done before this test is started. At all tests (where `diskquota_test`
is used) command `CREATE EXTENSION diskquota_test;` is executed.
```

red1452 committed:
fixes at readme
